# **Used Car Listings on eBay Kleinanzeigen**





## **Objectives**


In this guided project, we'll work with a dataset of used cars from eBay Kleinanzeigen, a classifieds section of the German eBay website.

The dataset was originally scraped and uploaded to Kaggle by user *orgesleka*.
The original dataset isn't available on Kaggle anymore, but you can find it <a href='https://data.world/data-society/used-cars-data'>here</a>.

DataQeust has made the following modification on the original data:

* sampled 50,000 data points from the full dataset, to ensure code runs quicker
* dirtied the dataset a bit to more closely resemble what one would expect from a scraped dataset (the version uploaded to Kaggle was cleaned to be easier to work with)

The data dictionary provided with data is as follows:

* dateCrawled - When this ad was first crawled. All field-values are taken from this date.
* name - Name of the car.
* seller - Whether the seller is private or a dealer.
* offerType - The type of listing
* price - The price on the ad to sell the car.
* abtest - Whether the listing is included in an A/B test.
* vehicleType - The vehicle Type.
* yearOfRegistration - The year in which which year the car was first registered.
* gearbox - The transmission type.
* powerPS - The power of the car in PS.
* model - The car model name.
* kilometer - How many kilometers the car has driven.
* monthOfRegistration - The month in which which year the car was first registered.
* fuelType - What type of fuel the car uses.
* brand - The brand of the car.
* notRepairedDamage - If the car has a damage which is not yet repaired.
* dateCreated - The date on which the eBay listing was created.
* nrOfPictures - The number of pictures in the ad.
* postalCode - The postal code for the location of the vehicle.
* lastSeenOnline - When the crawler saw this ad last online.

The aim of this project is to clean the data and analyze the included used car listings.

In [1]:
# import the libraries needed to read the dataset
import pandas as pd 
import numpy 

# read the autos.csv CSV file into pandas
autos=pd.read_csv('autos.csv')

In [2]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_T�V_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___T�V_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [3]:
autos.info() 
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_T�V_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

Columns that includes null data(areas that may need data cleaning): 
* vehicleType
* gearbox
* model
* fuelType
* notRepairedDAmage

Also column in price are objects, these might need converting if we want to do analysis on pirce.


## **Cleaning column names**

The column names also use camecase instead of Python's typical snakecase, let's convert the names to snakecase so we can replace spaces with underscores, and reword some names based on the data dictionary to be more descriptive.

In [4]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [5]:
# column names that we reworded or changed to snakecase
corrected_names={'dateCrawled':'date_crawled','name':'name', 'seller':'seller','offerType':'offer_type',
                  'price':'price', 'abtest':'abtest','vehicleType':'vehicle_type','yearOfRegistration':'registration_year',
                  'gearbox': 'gearbox','powerPS':'powerPS', 'model': 'model','odometer':'odometer',
                  'monthOfRegistration':'registration_month', 'fuelType':'fuel_type','brand':'brand',
                  'notRepairedDamage':'unrepaired_damage','dateCreated':'ad_created', 'nrOfPictures':'num_pictures', 
                  'postalCode':'postal_code','lastSeen':'last_seen'}

# assign the modified column names back to original DataFrame.columns attribute
autos.columns=autos.columns.map(corrected_names)

# check to see our modified column names worked
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_T�V_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  powerPS   model   odometer  registration_month fuel_type  \
0    manuell      158  andere  150,000km                   3       lpg   
1  automatik      286     7er  150,000km                   6    benzin   
2    manuell      102    golf   70,000km                   7    benzin   
3  automatik       71  fortwo   70,000km                   6    benzin   
4    manuell        0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  num_pictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

Let's do some data exploration to see what other cleaning tasks needs to be done. We will look for:
* Text columns where all or almost all values are the same, since they don't have useful information for analysis, they can be dropped
* Clean and convert examples of numeric data stored as text

In [6]:
# DataFrame.describe() to help use look at descriptive statistics for all columns
autos.describe(include='all') #include='all' to get both categorical and numeric columns

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-11 22:38:16  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox       powerPS  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  num_pictures   postal_code            last_seen  
count                 50000       50000.0  50000.000000                50000  
unique                   76           NaN           NaN                39481  
top     2016-04-03 00:00:00           NaN           NaN  2016-04-07 06:17:27  
freq                   1946           NaN           NaN                    8  
mean                    NaN           0.0  50813.627300                  NaN  
std                     NaN           0.0  25779.747957                  NaN  
min                     NaN           0.0   1067.000000                  NaN  
25%                     NaN           0.0  30451.000000                  NaN  
50%                     NaN           0.0  49577.000000                  NaN  
75%                     NaN           0.0  71540.000000                  NaN  
max                     NaN           0.0  99998.000000                  NaN

In [7]:
autos['model'].value_counts().head(5)

golf      4024
andere    3528
3er       2761
polo      1757
corsa     1735
Name: model, dtype: int64

In [8]:
autos['unrepaired_damage'].value_counts().head(5)

nein    35232
ja       4939
Name: unrepaired_damage, dtype: int64

The columns that mostly contains one value that doesn't seems to provide much information for us are:
* num_pictures: where there are 0 pictures
* seller & offer_type: almost all the values are the same

The columns that need more investigation are:
* price: there are 1421 cases where the price of the car is $0




In [9]:
autos['num_pictures'].value_counts()


0    50000
Name: num_pictures, dtype: int64

We will drop the 3 columns that doesn't seems to provide us much information

In [10]:
autos=autos.drop(['num_pictures','seller','offer_type'], axis=1)


In [11]:
autos.columns

Index(['date_crawled', 'name', 'price', 'abtest', 'vehicle_type',
       'registration_year', 'gearbox', 'powerPS', 'model', 'odometer',
       'registration_month', 'fuel_type', 'brand', 'unrepaired_damage',
       'ad_created', 'postal_code', 'last_seen'],
      dtype='object')


The columns that contains numeric data stored as text that needs to be cleaned:
* price
* odometer
* All the columns that contains a date



## **Convert Column *price* and *odometer* to Numeric dtype**

* Remove any non-numeric characters.
* Convert the column to a numeric dtype.


In [12]:
# let's identify some of the patterns and special cases of non-numeric characters
# in price and odometer columns that we want to remove 
autos['price'].unique()


array(['$5,000', '$8,500', '$8,990', ..., '$385', '$22,200', '$16,995'],
      dtype=object)

In [13]:
autos['odometer'].unique()

array(['150,000km', '70,000km', '50,000km', '80,000km', '10,000km',
       '30,000km', '125,000km', '90,000km', '20,000km', '60,000km',
       '5,000km', '100,000km', '40,000km'], dtype=object)

In [14]:
# remove non-numeric characters 

# current version of DataFrame.str.replace uses regular expressions as default, 
# so we have to set regex to false, also have to set DataFrame.astype(str) to turn 
# series class into string class

autos['price']=(autos['price']
                .astype(str)
                .str.replace('$','',regex=False)
                .str.replace(',','')
                .astype(int)
               )


autos['odometer']=(autos['odometer']
                   .astype(str)
                   .str.replace('km','',regex=False)
                   .str.replace(',','')
                   .astype(int)
                  )



print(autos['price'].head())
print(autos['odometer'].head())



0    5000
1    8500
2    8990
3    4350
4    1350
Name: price, dtype: int64
0    150000
1    150000
2     70000
3     70000
4    150000
Name: odometer, dtype: int64


In [15]:
# rename the column odometer to odometer_km
autos.rename({'odometer':'odometer_km'},axis=1,inplace=True)

autos.head(1)


date_crawled                              name  price   abtest  \
0  2016-03-26 17:47:46  Peugeot_807_160_NAVTECH_ON_BOARD   5000  control   

  vehicle_type  registration_year  gearbox  powerPS   model  odometer_km  \
0          bus               2004  manuell      158  andere       150000   

   registration_month fuel_type    brand unrepaired_damage  \
0                   3       lpg  peugeot              nein   

            ad_created  postal_code            last_seen  
0  2016-03-26 00:00:00        79588  2016-04-06 06:45:54

## **Continue Exploring Odometer and Price**


Next, let's see if there are any values that look unrealistically high or low that we might want to remove for odometer_km and price columns

In [16]:
# to see how many unique values are in price column
print(autos['price'].unique().shape)
print('\n')

# to see the min/max/median/mean
print(autos['price'].describe())
print('\n')


print(autos['price'].value_counts().head(10))
print(autos['price'].value_counts().sort_index(ascending=False).head(20))


(2357,)


count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64


0       1421
500      781
1500     734
2500     643
1000     639
1200     639
600      531
800      498
3500     498
2000     460
Name: price, dtype: int64
99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
Name: price, dtype: int64


There are *1421* cases of \\$0 for the price of a car, also *11* cases where the price of car exceed \\$1,000,000, with *5* cars over \\$10,000,000. The most expensive car is $1 under \\$100,000,000.

In [17]:
autos["price"].value_counts().sort_index(ascending=True).head(25)

0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
13       2
14       1
15       2
17       3
18       1
20       4
25       5
29       1
30       7
35       1
40       6
45       4
47       1
49       4
50      49
Name: price, dtype: int64

There are quite a few sub \\$100 listings, however, since eBay is an auction site, opening bid at \\$1 could be a legitimate listing. 

As a result, we will keep the \\$1 items, but remove anything abouve \\$350,000, because the next price after \\$350,000 is \\$999,990, this seems like quite an unrealistic jump in numbers.

In [18]:
# we will only keep prices between $1-$350,000
autos=autos[autos['price'].between(1,350001)]
autos['price'].describe()

count     48565.000000
mean       5888.935591
std        9059.854754
min           1.000000
25%        1200.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price, dtype: float64

In [19]:
# to see how many unique values are in odometer_km column
print(autos['odometer_km'].unique().shape)
print('\n')

# to see the min/max/median/mean
print(autos['odometer_km'].describe())
print('\n')


print(autos['odometer_km'].value_counts().head(10))
print(autos['odometer_km'].value_counts().sort_index(ascending=True).head(10))


(13,)


count     48565.000000
mean     125770.101925
std       39788.636804
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64


150000    31414
125000     5057
100000     2115
90000      1734
80000      1415
70000      1217
60000      1155
50000      1012
5000        836
40000       815
Name: odometer_km, dtype: int64
5000      836
10000     253
20000     762
30000     780
40000     815
50000    1012
60000    1155
70000    1217
80000    1415
90000    1734
Name: odometer_km, dtype: int64


As we can see from the result, roughly 65% of the cars have over 150,000km mileage in its odometer, with the average odometer mileage of 125,732.7km.

Also note that the from 50% percentiles onward to max, mileage is a fixed 150,000km, add to the fact that the mileages are rounded, which might indicate that the seller had to choose from pre-set options for this field with the highest option been 150,000km.

## **Exploring the Date Columns**

Let's take a look at the date columns, we will differentiate the dates created by the crwaler and those from teh website itself:

- `date_crawled`: added by the crawler
- `last_seen`: added by the crawler
- `ad_created`: from the website
- `registration_month`: from the website
- `registration_year`: from the website

`date_crawled`, `last_seen`, and `ad_created` columns are all identified as string values by pandas.

In [20]:
# find out how the values in the threee string columns are formatted
autos[['date_crawled','ad_created','last_seen']][:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

In [21]:
# select the first 10 characters in each column that represent the day in the date
day_date_crawled=autos['date_crawled'].str[:10]
day_ad_created=autos['ad_created'].str[:10]
day_last_seen=autos['last_seen'].str[:10]

# chronological order of the distribution of values in date_crawled in percentages
(day_date_crawled
 .value_counts(normalize=True,dropna=False)
 .sort_index()
)



2016-03-05    0.025327
2016-03-06    0.014043
2016-03-07    0.036014
2016-03-08    0.033296
2016-03-09    0.033090
2016-03-10    0.032184
2016-03-11    0.032575
2016-03-12    0.036920
2016-03-13    0.015670
2016-03-14    0.036549
2016-03-15    0.034284
2016-03-16    0.029610
2016-03-17    0.031628
2016-03-18    0.012911
2016-03-19    0.034778
2016-03-20    0.037887
2016-03-21    0.037373
2016-03-22    0.032987
2016-03-23    0.032225
2016-03-24    0.029342
2016-03-25    0.031607
2016-03-26    0.032204
2016-03-27    0.031092
2016-03-28    0.034860
2016-03-29    0.034099
2016-03-30    0.033687
2016-03-31    0.031834
2016-04-01    0.033687
2016-04-02    0.035478
2016-04-03    0.038608
2016-04-04    0.036487
2016-04-05    0.013096
2016-04-06    0.003171
2016-04-07    0.001400
Name: date_crawled, dtype: float64

In [22]:
# ascending order of date_crawled by distribution percentages
(day_date_crawled
 .value_counts(normalize=True,dropna=False)
 .sort_values()
)


2016-04-07    0.001400
2016-04-06    0.003171
2016-03-18    0.012911
2016-04-05    0.013096
2016-03-06    0.014043
2016-03-13    0.015670
2016-03-05    0.025327
2016-03-24    0.029342
2016-03-16    0.029610
2016-03-27    0.031092
2016-03-25    0.031607
2016-03-17    0.031628
2016-03-31    0.031834
2016-03-10    0.032184
2016-03-26    0.032204
2016-03-23    0.032225
2016-03-11    0.032575
2016-03-22    0.032987
2016-03-09    0.033090
2016-03-08    0.033296
2016-04-01    0.033687
2016-03-30    0.033687
2016-03-29    0.034099
2016-03-15    0.034284
2016-03-19    0.034778
2016-03-28    0.034860
2016-04-02    0.035478
2016-03-07    0.036014
2016-04-04    0.036487
2016-03-14    0.036549
2016-03-12    0.036920
2016-03-21    0.037373
2016-03-20    0.037887
2016-04-03    0.038608
Name: date_crawled, dtype: float64

The crawled time period range over a one month period in March and April in 2016, distribution of listing is around 3%.

In [23]:
# chronological order of the distribution of values in last_seen in percentages
(day_last_seen
        .value_counts(normalize=True, dropna=False)
        .sort_index() 
        )

2016-03-05    0.001071
2016-03-06    0.004324
2016-03-07    0.005395
2016-03-08    0.007413
2016-03-09    0.009595
2016-03-10    0.010666
2016-03-11    0.012375
2016-03-12    0.023783
2016-03-13    0.008895
2016-03-14    0.012602
2016-03-15    0.015876
2016-03-16    0.016452
2016-03-17    0.028086
2016-03-18    0.007351
2016-03-19    0.015834
2016-03-20    0.020653
2016-03-21    0.020632
2016-03-22    0.021373
2016-03-23    0.018532
2016-03-24    0.019767
2016-03-25    0.019211
2016-03-26    0.016802
2016-03-27    0.015649
2016-03-28    0.020859
2016-03-29    0.022341
2016-03-30    0.024771
2016-03-31    0.023783
2016-04-01    0.022794
2016-04-02    0.024915
2016-04-03    0.025203
2016-04-04    0.024483
2016-04-05    0.124761
2016-04-06    0.221806
2016-04-07    0.131947
Name: last_seen, dtype: float64

`last_seen` records the date the crawler last saw any listing, which indicates the day a listing was removed, presumably because the car was sold.

We see a huge jump in percentage from values in `last_seen` in the last 3 days, which are 6-10 times the values from the rest of the month. The cause of such jump in values are probably not due to large deals of sold cars, it is more likely that these values are to do with the crawling period ending and crawler stopped indicating car sales.

In [24]:
# the numbers of ads created in different dates
print(autos["ad_created"].str[:10].unique().shape)

# chronological order of the distribution of values in ad_created in percentages
(day_ad_created
        .value_counts(normalize=True, dropna=False)
        .sort_index()
        )

(76,)


2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
                ...   
2016-04-03    0.038855
2016-04-04    0.036858
2016-04-05    0.011819
2016-04-06    0.003253
2016-04-07    0.001256
Name: ad_created, Length: 76, dtype: float64

In [25]:
# distribution of registration_year
autos['registration_year'].describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

## **Incorrect Registration Year Data**


Some numbers in the above data set are obviously wrong, with minimum value at 1000, 886 years before the birht of the first car, and the maximum is 9999, which is 7,983 years from 2016.

A car can't be registered after the listing was seen, any vehicle with a registration year above 2016 is definitely inaccurate. Earliest registration could be anywhere from the 1900s. 

In [26]:
# find out what percentage of the data has values outside 1900-2016
(~autos['registration_year'].between(1900,2016)).sum()/autos.shape[0]

0.038793369710697

Only less than 4% of the rows is outside the range of 1900-2016, it should be save to remove them.

In [27]:
# only keep the rows in the dataframe that is in the range  1900-2016
autos=autos[autos['registration_year'].between(1900,2016)]

autos['registration_year'].value_counts(normalize=True).head(20)

2000    0.067608
2005    0.062895
1999    0.062060
2004    0.057904
2003    0.057818
2006    0.057197
2001    0.056468
2002    0.053255
1998    0.050620
2007    0.048778
2008    0.047450
2009    0.044665
1997    0.041794
2011    0.034768
2010    0.034040
1996    0.029412
2012    0.028063
1995    0.026285
2016    0.026135
2013    0.017202
Name: registration_year, dtype: float64

Most cars are from the last 20 years.


## **Exploring the Prices on Brands**



In [28]:
# select the top 20 brand based on percentage of distribution 
autos['brand'].value_counts(normalize=True).head(20)

volkswagen        0.211264
bmw               0.110045
opel              0.107581
mercedes_benz     0.096463
audi              0.086566
ford              0.069900
renault           0.047150
peugeot           0.029841
fiat              0.025642
seat              0.018273
skoda             0.016409
nissan            0.015274
mazda             0.015188
smart             0.014160
citroen           0.014010
toyota            0.012703
hyundai           0.010025
sonstige_autos    0.009811
volvo             0.009147
mini              0.008762
Name: brand, dtype: float64

Perhaps not coming as a surprise to anyone, the top five brands in the German eBay website are all from German manufactureers.

Let's rule out brands representing less than 5% of total listings.

In [29]:
# assign brand_counts to percentage of distribution from largest to lowest
brand_counts=autos['brand'].value_counts(normalize=True)

# filter the rows that meet the boolean condition of brand_counts larger than 5%
# use Series.index attribute to access the labels
brands_wewant=brand_counts[brand_counts>.05].index

brands_wewant


Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object')

In [34]:
#empty dictionary to hold the aggregate data
mean_price={}
#we will assign brand name as key, and mean price as value
for brand in brands_wewant:
    brand_same=autos[autos['brand']==brand]
    
    price=brand_same['price'].mean()
    mean_price[brand]=int(price)
    


mean_price

{'volkswagen': 5402,
 'bmw': 8332,
 'opel': 2975,
 'mercedes_benz': 8628,
 'audi': 9336,
 'ford': 3749}

BMW, Mercedes Benz, and Audi are significantly more expensive than the other brands. Opel and Ford are less expensive.

Let's find the average mileage for these brands and see if there's visible link with mean price.

In [37]:
#create an empty dictionary to store mean mileage
mean_mileage={}
#we will assign brand name as key, and mean mileage as value
for brand in brands_wewant:
    brand_same=autos[autos['brand']==brand]
    
    mile=brand_same['odometer_km'].mean()
    mean_mileage[brand]=int(mile)
mean_mileage

{'volkswagen': 128707,
 'bmw': 132572,
 'opel': 129310,
 'mercedes_benz': 130788,
 'audi': 129157,
 'ford': 124266}

In [58]:
#convert the mean_price dictionary to series objects using series constructor
bmp_series=pd.Series(mean_price).sort_values(ascending=False)
print(bmp_series)
print('\n')

#convert the mean_mileage dictionary to series objects using series constructor
bmm_series=pd.Series(mean_mileage).sort_values(ascending=False)
print(bmm_series)

audi             9336
mercedes_benz    8628
bmw              8332
volkswagen       5402
ford             3749
opel             2975
dtype: int64


bmw              132572
mercedes_benz    130788
opel             129310
audi             129157
volkswagen       128707
ford             124266
dtype: int64


In [59]:
#create a dataframe from the two series object using the dataframe constructor
brands_mean=pd.DataFrame(bmp_series,columns=['mean_price'])
brands_mean['mean_mileage']=bmm_series

brands_mean

mean_price  mean_mileage
audi                 9336        129157
mercedes_benz        8628        130788
bmw                  8332        132572
volkswagen           5402        128707
ford                 3749        124266
opel                 2975        129310

Interestingly, Mercedes Benz have almost 1.6 times the average price compare to Volkswagen and Audi have more than 1.7 times the average price compare to Volkswagen, while both brands carry more average mileage over Vokkswagen, Mercedes Benz 1.6% more average mileage and Audi 0.35% more average mileage over Volkswagen.

Also Opel have the lowest average price while have more average mileage over Ford, Volkswagen, and Audi.

Ford been the 2nd lowest average price brand while having the least average mileage.

## **Conclusion**

It does seems like there is a small pattern, if not only exists in the top 6 brands we analyzed, that the more expensive the car is, the higher the average mileage it carries. And the price slightly reflects more on the brand, not as much on mileage.